meta/sai_rpc_frontend.cpp

```diff
@@ -0,0 +1,870 @@
+/**
+ * Copyright (c) 2021 Microsoft Open Technologies, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *    not use this file except in compliance with the License. You may obtain
+ *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    sai_rpc_frontend.cpp
+ *
+ * @brief   This module contains RPC server handler and helper functions
+ */
+
+#include <arpa/inet.h>
+
+#include <iostream>
+#include <cstring>
+#include "sai_rpc.h"
+
+extern "C" {
+#include "sai.h"
+#include "saitypes.h"
+#include "saimetadatatypes.h"
+#include "saimetadatautils.h"
+#include "saimetadata.h"
+}
+
+using namespace ::sai;
+
+unsigned int sai_thrift_mac_t_parse(const std::string s, void *data) {
+  unsigned int i, j = 0;
+  unsigned char *m = static_cast<unsigned char *>(data);
+  memset(m, 0, 6);
+  for (i = 0; i < s.size(); i++) {
+    char let = s.c_str()[i];
+    if (let >= '0' && let <= '9') {
+      m[j / 2] = (m[j / 2] << 4) + (let - '0');
+      j++;
+    } else if (let >= 'a' && let <= 'f') {
+      m[j / 2] = (m[j / 2] << 4) + (let - 'a' + 10);
+      j++;
+    } else if (let >= 'A' && let <= 'F') {
+      m[j / 2] = (m[j / 2] << 4) + (let - 'A' + 10);
+      j++;
+    }
+  }
+  return (j == 12);
+}
+
+void sai_thrift_ip4_t_parse(const std::string s, unsigned int *m) {
```

Comment:
How about adding unit tests for the APIs performing SAI thrift to SAI and vice versa.